### PR TITLE
Teal stateless

### PR DESCRIPTION
**Pull request by algobolson**
make TEAL stateless and cacheable
no longer has direct access to Round or TimeStamp (instead has FirstValidTime)
`rand` removed until it can be done right
add LogicSig Eval caching to TransactionPool
better docs

### CODE REVIEW

File: ledger/eval.go [hunk 1]
```diff
@@ -479,39 +484,32 @@ func (eval *BlockEvaluator) transaction(txn transactions.SignedTxn, ad transacti
...
+			err = eval.checkLogicSig(txn, txgroup, groupIndex)
 			if err != nil {
-				return fmt.Errorf("transaction %v: rejected by logic err=%s", txn.ID(), err)
-			}
-			if !pass {
-				return fmt.Errorf("transaction %v: rejected by logic", txn.ID())
+				eval.txcache.EvalRemember(txn, err.Error())
+				return err
 			}
+			eval.txcache.EvalRemember(txn, "")
```
Comment (pzbitskiy) [inline ledger/eval.go]: eval.txcache.EvalRemember(txn, nil) ?

File: data/transactions/logic/globalfield_string.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+	// An "invalid array index" compiler error signifies that the constant values have changed.
+	// Re-run the stringer command to generate them again.
+	var x [1]struct{}
+	_ = x[MinTxnFee-0]
+	_ = x[MinBalance-1]
+	_ = x[MaxTxnLife-2]
+	_ = x[ZeroAddress-3]
+	_ = x[GroupSize-4]
+	_ = x[invalidGlobalField-5]
+}
```
Comment (pzbitskiy) [inline data/transactions/logic/globalfield_string.go]: what kind of magic is it?

File: ledger/eval.go (hunk 1 above)
Comment (pzbitskiy) [reply ledger/eval.go]: Nevermind,  nil can't be used as a value for a string field.

File: data/transactions/logic/globalfield_string.go (hunk 2 above)
Comment (algobolson) [reply data/transactions/logic/globalfield_string.go]: x has one element. x[0] is the only valid index. if any of the constants change, this generated code will break signaling that it needs to be re-generated. (I'd normally not check in generated code, but we're already doing that elsewhere with this same `stringer` generator)

File: data/transactions/logic/eval.go [hunk 3]
```diff
@@ -1021,57 +996,59 @@ func opDup(cx *evalContext) {
...
+	case GroupIndex:
 		sv.Uint = uint64(cx.GroupIndex)
-	case 21:
+	case TxID:
 		txid := txn.ID()
 		sv.Bytes = txid[:]
-	case 22:
-		sv.Uint = cx.GroupSenders[cx.GroupIndex].MicroAlgos.Raw
-	case 23:
+	/*case 22: // SenderBalance
```
Comment (justicz) [inline data/transactions/logic/eval.go]: are you planning on leaving these comments in?

File: data/pools/transactionPool.go [hunk 4]
```diff
@@ -72,6 +76,7 @@ func MakeTransactionPool(ledger *ledger.Ledger, transactionPoolStatusSize int, l
 		ledger:          ledger,
 		statusCache:     makeStatusCache(transactionPoolStatusSize),
```
Comment (tsachiherman) [inline data/pools/transactionPool.go]: The `statusCache` contains only the *failed* transaction groups. Shouldn't we rename it ?

File: data/pools/transactionPool.go [hunk 5]
```diff
@@ -337,6 +342,20 @@ func (pool *TransactionPool) Verified(txn transactions.SignedTxn) bool {
...
+func (pool *TransactionPool) EvalOk(txid transactions.Txid) (tx transactions.SignedTxn, txErr string, found bool) {
+	pool.lcmu.RLock()
+	defer pool.lcmu.RUnlock()
+	return pool.lsigCache.check(txid)
+}
+
+// EvalRemember sets an error string from LogicSig Eval for some SignedTxn
+func (pool *TransactionPool) EvalRemember(tx transactions.SignedTxn, errString string) {
+	pool.lcmu.Lock()
+	defer pool.lcmu.Unlock()
```
Comment (tsachiherman) [inline data/pools/transactionPool.go]: I like the way you wrapped the lock around the object.

File: data/pools/transactionPool.go [hunk 6]
```diff
@@ -337,6 +342,20 @@ func (pool *TransactionPool) Verified(txn transactions.SignedTxn) bool {
 	return pendingSigTxn.Sig == txn.Sig && pendingSigTxn.Msig.Equal(txn.Msig) && pendingSigTxn.Lsig.Equal(&txn.Lsig)
 }
 
+// EvalOk for LogicSig Eval of a txn by txid, returns the SignedTxn, error string, and found.
+func (pool *TransactionPool) EvalOk(txid transactions.Txid) (tx transactions.SignedTxn, txErr string, found bool) {
```
Comment (tsachiherman) [inline data/pools/transactionPool.go]: nit: The first return value is not being used, and can be omitted from the function signature; also, the second argument need to return just whether it's != "" or not. ( i.e. error'ed or not ).

File: data/transactions/logic/eval.go (hunk 3 above)
Comment (algobolson) [reply data/transactions/logic/eval.go]: cleaned up

File: data/pools/transactionPool.go (hunk 4 above)
Comment (algobolson) [reply data/pools/transactionPool.go]: Yeah, it is that.
```
// failedTxnStatuses records failure messages from recently failed transactions, indexed by Txid.
// This allows algod clients to find out why the transaction they submitted failed.
failedTxnStatuses: makeStatusCache(...)
```

File: data/pools/transactionPool.go (hunk 6 above)
Comment (algobolson) [reply data/pools/transactionPool.go]: dropped the unused argument, but there's one place we return the error string if it's not ""

**Comment by tsachiherman** [comment]
@algobolson - could you take another look and update the code so it would compile without any errors ? thanks!

**Pull request merged**